### PR TITLE
Add database injection

### LINKED ISSUE #986: Add database injection to selenium tests
Issue raised by: [developer]

### What?
We need to add database injection strategy to selenium tests. This will allows us to insert runs into the database and variables to make sure they are all correctly loaded into the database. An alternative would be to test the values that are pre populated in the setup, but this is limiting.

### Where?
`webtests/`

How to test the issue is resolved
Deeper testing should be possible

### PR DESCRIPTION
**Pull request by keiranjprice101**
### Summary of work
Majority of commits add a way to inject data before the action runs. An explanation of this is within the README commit 8755ac4

 - 6b092e2 This commit relates to issue #1036
 - 5fc3f06  is also related to #1036 as most data will be coming from the database now
 - 1adc97f This commit turns off duplication search within pylint due to false positives.

### Additional comments
 

closes #986

**Before merging ensure the release notes have been updated**

### CODE REVIEW

File: webtests/database.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    Return a PyMySql connection object for querying the database
+    :returns: (Connection) The PyMySql Connection object.
+
+    """
+    return pymysql.connect(host=get_str("DATABASE", "host"),
+                           user=get_str("DATABASE", "user"),
+                           password=get_str("DATABASE", "password"),
+                           database=get_str("DATABASE", "name"),
+                           charset='utf8mb4',
+                           cursorclass=pymysql.cursors.DictCursor)
```
Comment (dtasev) [inline webtests/database.py:34]: Can this use Django's ORM?

File: webtests/database.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    return rows
+
+
+def inject_dataset(dataset: Path) -> None:
+    """
+    Given the filename of an included dataset, inject the values into the database.
+    :param dataset: (Path) Path object for dataset
+    """
+    dataset = parse_dataset(dataset)
+    queries = ["INSERT INTO %s (%s) VALUES (%s)" %
```
Comment (dtasev) [inline webtests/database.py:60]: Can we avoid any raw SQL and go through the ORM?

**Review by dtasev — CHANGES_REQUESTED** [review]
Can we not get this through the Django ORM? The injection here seems to be using pymysql to directly insert into the DB, but that won't work for anything but MySQL. 

With https://github.com/ISISScientificComputing/autoreduce/issues/1072 I'm removing use of SQLAlchemy, and then we'll be using the Django ORM everywhere, which makes it super easy to switch backends (e.g. to SQLite3 for fast testing & local running).

**Comment by keiranjprice101** [comment]
@dtasev 
Tests now use `StaticLiveServerTestCase` which self deploys. Data now comes from fixtures. Also changed `url` to `re_path` as `url` is deprecated.

Updated README to reflect to instructions. 

Best to check files changed not commits. I had a bit of a dull moment with a merge conflict.

**Comment by dtasev** [comment]
You'll have to rebase again - a lot of these urls don't exist anymore, they are now included from `instrument.urls` (all `instrument/`) and `reduction_viewer.urls` (most `runs/`)

File: WebApp/autoreduce_webapp/selenium_tests/pages/instrument_summary_page.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        :param version: (int) the version of the run to click
+        :return: (RunSummaryPage) The page object of the opened run summary page.
+        """
+        runs = self.driver.find_elements_by_class_name("run-num-links")
+        run_string = f"{run_number} - {version}" if version else f"{run_number}"
+        for run in runs:
+            if run.text == run_string:
+                run.click()
+                return RunSummaryPage(self.driver, self.instrument, run_number, version)
+        raise NoSuchElementException
```
Comment (dtasev) [inline WebApp/autoreduce_webapp/selenium_tests/pages/instrument_summary_page.py:68]: That seems like a bad merge - will fix

File: WebApp/autoreduce_webapp/autoreduce_webapp/settings.py [hunk 4]
```diff
@@ -105,24 +106,27 @@ def get_str(section, key):
...
-    #         'HOST': get_str('DATABASE', 'host'),
-    #         'PORT': get_str('DATABASE', 'port'),
-    #         'OPTIONS': {
-    #             'init_command': "SET sql_mode='STRICT_TRANS_TABLES'",
-    #         },
-    #     }
-    'default': {
-        'ENGINE': 'django.db.backends.sqlite3',  # Add 'postgresql_psycopg2', 'mysql', 'sqlite3' or 'oracle'.
-        'NAME': f'{PROJECT_ROOT}/sqlite3.db',  # Or path to database file if using sqlite3.
+if 'test' in sys.argv:
```
Comment (keiranjprice101) [inline WebApp/autoreduce_webapp/autoreduce_webapp/settings.py:109]: Needs changing

File: .github/workflows/run-tests.yml [hunk 5]
```diff
@@ -46,6 +46,19 @@ jobs:
   selenium:
     name: Selenium
     runs-on: ubuntu-18.04
+    services:
```
Comment (keiranjprice101) [inline .github/workflows/run-tests.yml:49]: Needs changing

File: WebApp/autoreduce_webapp/selenium_tests/configuration.py [hunk 6]
```diff
@@ -15,8 +15,8 @@
 
 from utils.project.structure import get_project_root
 
-CONFIG_PATH = Path(get_project_root(), "webtests/config.json")
-TEMP_CONFIG_PATH = Path(get_project_root(), "webtests/temp_config.json")
+CONFIG_PATH = Path(get_project_root(), "WebApp/autoreduce_webapp/selenium_tests/config.json")
+TEMP_CONFIG_PATH = Path(get_project_root(), "Webapp/autoreduce_webapp/selenium_tests/temp_config.json")
```
Comment (dtasev) [inline WebApp/autoreduce_webapp/selenium_tests/configuration.py:19]: Should use `PROJECT_ROOT` with `from utils.project.structure import PROJECT_ROOT`

**Review by dtasev — CHANGES_REQUESTED** [review]

File: WebApp/autoreduce_webapp/selenium_tests/configuration.py (hunk 6 above)
Comment (keiranjprice101) [reply WebApp/autoreduce_webapp/selenium_tests/configuration.py:19]: Aye, this wasn't available originally. Will add it Cheers 👍

**Comment by dtasev** [comment]
The fail to find All instrument, failed jobs etc (from https://github.com/ISISScientificComputing/autoreduce/pull/1044/checks?check_run_id=1917570724) makes me think I might have messed up something with the login in the tests - you don't see those buttons unless logged in

**Comment by dtasev** [comment]
Can we remove the requirement to have to manually create a config.json? It seems like it gets overwritten for every test anyway so it doesn't seem necessary anymore

**Review by dtasev — CHANGES_REQUESTED** [review]
~~The latest changes fixed running headless for me, but if I switch headless to false I get:~~

```
selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: exited abnormally.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/google-chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

~~I think it'd be useful having the option to not run headless, especially during writing tests, but also for debugging. Any way we can have both?~~

**Comment by dtasev** [comment]
Nevermind that one's on me - I was running it from SSH but hadn't set the DISPLAY env variable. Actually that might be a good check to have in `get_chrome_driver`!

**Review by dtasev — APPROVED** [review]
Looks good, runs fine locally both headless and not.

**Pull request merged**